cleanrl/ppo_mujoco_envpool_xla_jax.py

```diff
@@ -0,0 +1,679 @@
+        next_obs,
+        next_done,
+        next_truncated,
+        key,
+        handle,
+        step_once_fn,
+        max_steps,
+    ):
+        (agent_state, episode_stats, next_obs, next_done, next_truncated, key, handle,), storage = jax.lax.scan(
+            step_once_fn,
+            (
+                agent_state,
+                episode_stats,
+                next_obs,
+                next_done,
+                next_truncated,
+                key,
+                handle,
+            ),
+            (),
+            max_steps,
+        )
+        return (
+            agent_state,
+            episode_stats,
+            next_obs,
+            next_done,
+            next_truncated,
+            storage,
+            key,
+            handle,
+        )
+
+    rollout = partial(
+        rollout,
+        step_once_fn=partial(step_once, env_step_fn=step_env_wrappeed),
+        max_steps=args.num_steps,
+    )
+    for update in range(1, args.num_updates + 1):
+        update_time_start = time.time()
+        (
+            agent_state,
+            episode_stats,
+            next_obs,
+            next_done,
+            next_truncated,
+            storage,
+            key,
+            handle,
+        ) = rollout(agent_state, episode_stats, next_obs, next_done, next_truncated, key, handle)
+        global_step += args.num_steps * args.num_envs
+        agent_state, loss, pg_loss, v_loss, entropy_loss, approx_kl, key = update_ppo(
+            agent_state,
+            next_obs,
+            next_done,
+            next_truncated,
+            storage,
+            key,
+        )
+        print(f"pg_loss={pg_loss.mean()}, loss={loss.mean()}, v_loss={v_loss.mean()}, entropy_loss={entropy_loss.mean()}")
+        avg_episodic_return = np.mean(jax.device_get(episode_stats.returned_episode_returns))
+        avg_episodic_length = np.mean(jax.device_get(episode_stats.returned_episode_lengths))
+        print(
+            f"global_step={global_step}, avg_episodic_length={avg_episodic_length}, avg_episodic_return={avg_episodic_return}, SPS_update={int(args.num_steps * args.num_envs / (time.time() - update_time_start))}"
+        )
+
+        # TRY NOT TO MODIFY: record rewards for plotting purposes
+        writer.add_scalar("charts/avg_episodic_return", avg_episodic_return, global_step)
+        writer.add_scalar("charts/avg_episodic_length", avg_episodic_length, global_step)
+        writer.add_scalar("charts/learning_rate", agent_state.opt_state[1].hyperparams["learning_rate"].item(), global_step)
+        writer.add_scalar("losses/value_loss", v_loss.mean().item(), global_step)
+        writer.add_scalar("losses/policy_loss", pg_loss.mean().item(), global_step)
+        writer.add_scalar("losses/entropy", entropy_loss.mean().item(), global_step)
+        writer.add_scalar("losses/approx_kl", approx_kl.mean().item(), global_step)
+        writer.add_scalar("losses/loss", loss.mean().item(), global_step)
```

Comment:
All the logged values should use the value from the last minibatch as opposed to the mean (please see `ppo_atari_envpool_xla_jax_scan.py`). This deviation might cause some confusion as to why the logged values are not consistent with previous implementations.